File: Leetcode/recoverTree.java
TreeNode prev;
    TreeNode left;
    TreeNode right;
    public void recoverTree(TreeNode root) {
        if (root == null) {
            return;
        }
        prev = null;
        left = null;
        right = null;
        recoverTree_helper(root);
        if (left != null && right != null) {
            int temp = left.val;
            left.val = right.val;
            right.val = temp;
        }
    }
    public void recoverTree_helper(TreeNode root) {
        if (root == null) {
            return;
        }
        recoverTree_helper(root.left);
        if (prev != null && prev.val > root.val) {
            if (left == null) {
                left = prev;
            }
            right = root;
        }
        prev = root;
        recoverTree_helper(root.right);
    }

File: Leetcode/recoverTree.swift
func recoverTree(root: TreeNode?) {
        var prev: TreeNode?
        var left: TreeNode?
        var right: TreeNode?
        recoverTree(root, &prev, &left, &right);
        if left != nil && right != nil {
            var temp = left!.val;
            left!.val = right!.val
            right!.val = temp
        }
    }
    func recoverTree(root: TreeNode?, inout _ prev: TreeNode?, inout _ left: TreeNode?, inout _ right: TreeNode?) {
        if root == nil {
            return
        }
        
        recoverTree(root!.left, &prev, &left,  &right)
        if (prev != nil && prev!.val > root!.val) {
            if left == nil {
                left = prev!
            }
            right = root!
        }
        prev = root!
        recoverTree(root!.right, &prev, &left, &right)
    }
